File: code/scenes/src/main/java/org/fbme/scenes/controllers/SelectionModel.java
package org.fbme.scenes.controllers;

import org.jetbrains.annotations.NotNull;

import java.util.Set;

public interface SelectionModel<T> {

    Set<T> getSelectedComponents();

    void clear();

    void setSelected(T component, boolean selected);

    default boolean isSelected(T component) {
        return getSelectedComponents().contains(component);
    }

    void addListener(@NotNull Listener<T> listener);

    void removeListener(@NotNull Listener<T> listener);

    interface Listener<T> {
        void onComponentSelected(T component, boolean selected);
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/SelectionModel.kt
package org.fbme.scenes.controllers

interface SelectionModel<T> {
    val selectedComponents: Set<T>

    fun clear()
    fun setSelected(component: T, selected: Boolean)
    fun isSelected(component: T): Boolean {
        return selectedComponents.contains(component)
    }

    fun addListener(listener: Listener<T>)
    fun removeListener(listener: Listener<T>)

    interface Listener<T> {
        fun onComponentSelected(component: T, selected: Boolean)
    }
}
